### PR TITLE
refactor: Remove options generic parameter on OperationBase

### PR DESCRIPTION
**Pull request by nbbeeken**
Instead of maintaining the typing for bespoke options per operation,
each operation has the option properties assigned directly to itself.

NODE-2815

### Description

Each operation now `implements` its options interface to make sure we maintain the correct properties on the respective classes. There were two special cases in map_reduce and  create_collections where we used a blockList to omit properties from the command, with the new scheme it made more sense to make these into allowLists. 

I tried to minimize the impact outside of the operations subdirectory since Cursor is undergoing a number of changes, you'll notice a cast in CoreCursor and Cursor that blatantly overrides type warnings, it is my belief that we can remove this as part of the new cursor design work. 

Another exception to the options removal was the addition of the bsonOptions getter, this I think could be better but looking for feedback: the cursor looks for a `bsonOptions` property (separate from the options property) to pass through to getMore execution (tricky bug arose here where the first doc fetched by the cursor was deserialized with the user's options and the following one's were deserialized with the defaults). My solution was to make a bespoke bsonOptions object in a getter on OperationsBase despite all the options being defined on the operation directly. I could move this logic into the cursor function that fetches these options instead?

Lastly, and you'll notice a comment on it but the last line of executeCommands in `src/bulk/common.ts` we seem to swallow an error, maybe that is correct, but when fixing another bug I was having tests that had both an undefined `err` and undefined `result` which feels like odd behavior.

### CODE REVIEW

File: src/operations/add_user.ts [hunk 1]
```diff
@@ -18,11 +18,16 @@ export interface AddUserOptions extends CommandOperationOptions {
...
 
 /** @internal */
-export class AddUserOperation extends CommandOperation<AddUserOptions, Document> {
+export class AddUserOperation extends CommandOperation implements AddUserOptions {
   db: Db;
   username: string;
   password?: string;
 
+  /** @deprecated Please use db.command('createUser', ...) instead for this option */
+  digestPassword?: null;
```
Comment (reggi) [inline src/operations/add_user.ts:27]: Can you add the deprecated props to the bottom of this list? This can read as `digestPassword` is the only deprecated, or all that follow are as well.

File: src/operations/bulk_write.ts [hunk 2]
```diff
@@ -11,10 +17,21 @@ import type {
...
-export class BulkWriteOperation extends OperationBase<BulkWriteOptions, BulkWriteResult> {
+export class BulkWriteOperation
+  extends OperationBase<BulkWriteResult>
+  implements BulkWriteOptions, HasRetryableWrites, HasWriteConcern {
   collection: Collection;
   operations: AnyBulkWriteOperation[];
 
+  bypassDocumentValidation?: boolean;
+  ordered?: boolean;
+  /** @deprecated use `ordered` instead */
```
Comment (reggi) [inline src/operations/bulk_write.ts:28]: This one for `@deprecated` too.

File: src/operations/find_and_modify.ts [hunk 3]
```diff
@@ -33,21 +34,37 @@ export interface FindAndModifyOptions extends CommandOperationOptions {
...
   collection: Collection;
   query: Document;
   sort?: Sort;
   doc?: Document;
 
+  new?: boolean;
+  remove?: boolean;
+  upsert?: boolean;
+  projection?: Document;
+  /** @deprecated Use projection instead */
```
Comment (reggi) [inline src/operations/find_and_modify.ts:59]: Here too

File: src/operations/map_reduce.ts [hunk 4]
```diff
@@ -96,55 +103,52 @@ export class MapReduceOperation extends CommandOperation<MapReduceOptions, Docum
...
     }
 
     // Add any other options passed in
-    for (const n in options) {
-      // Only include if not in exclusion list
-      if (exclusionList.indexOf(n) === -1) {
-        mapCommandHash[n] = (options as any)[n];
+    for (const n in this) {
+      // Only include if defined on this
+      if (optionsAllowList.has(n)) {
```
Comment (reggi) [inline src/operations/map_reduce.ts:120]: The other ops that do this check for null values / functions wondering if there's a way we can gain consistency here?

File: src/operations/operation.ts [hunk 5]
```diff
@@ -34,23 +34,42 @@ export interface OperationOptions extends BSONSerializeOptions {
...
+    const bsonOptions: Document = {
+      promoteBuffers: this.promoteBuffers,
+      promoteValues: this.promoteValues,
+      promoteLongs: this.promoteLongs,
+      raw: this.raw,
+      ignoreUndefined: this.ignoreUndefined
+    } as const;
+
+    for (const key in bsonOptions) {
+      bsonOptions[key] === undefined && delete bsonOptions[key];
```
Comment (reggi) [inline src/operations/operation.ts:64]: Can we do this without using `delete`? I believe we try to stay away from it when possible.

Comment (reggi) [reply src/operations/operation.ts:64]: I do like this getter BTW 🤩

File: src/operations/replace_one.ts [hunk 6]
```diff
@@ -19,15 +19,16 @@ export interface ReplaceOptions extends CommandOperationOptions {
   /** When true, creates a new document if no document matches the query */
   upsert?: boolean;
 
-  // FIXME:
+  // FIXME: ????!!!! neal why?
```
Comment (reggi) [inline src/operations/replace_one.ts:22]: 😿

Comment (reggi) [reply src/operations/replace_one.ts:22]: Maybe the intention is to have a `replace` operation that does both, or a `replace_multi`?

**Review by reggi — APPROVED** [review]
LGTM, some little comments here and there but nothing stands out 🙇‍♂️

Really appreciate this work, it's one step closer toward getting rid of `.s` or `.options` and having properties attached to the class is much easier to reason about. Cleaner classes!

File: src/operations/create_collection.ts [hunk 7]
```diff
@@ -10,18 +10,19 @@ import type { PkFactory } from '../mongo_client';
 // eslint-disable-next-line
 import type { Collection } from '../collection';
 
-const ILLEGAL_COMMAND_FIELDS = new Set([
-  'w',
-  'wtimeout',
-  'j',
-  'fsync',
+const optionsAllowList = new Set([
```
Comment (emadum) [inline src/operations/create_collection.ts:13]:
```suggestion
const OPTIONS_ALLOW_LIST = new Set([
```

I think our convention is to format top-level constants in uppercased snake case.

File: src/operations/map_reduce.ts [hunk 8]
```diff
@@ -15,16 +15,17 @@ import type { Sort } from './find';
...
-  'readPreference',
-  'session',
-  'bypassDocumentValidation',
-  'w',
-  'wtimeout',
-  'j',
-  'writeConcern',
-  'scope' // this option is reformatted thus exclude the original
-];
+const optionsAllowList = new Set([
```
Comment (emadum) [inline src/operations/map_reduce.ts:18]:
```suggestion
const OPTIONS_ALLOW_LIST = new Set([
```

File: src/operations/map_reduce.ts (hunk 4 above)
Comment (emadum) [inline src/operations/map_reduce.ts:120]:
```suggestion
      if (OPTIONS_ALLOW_LIST.has(n)) {
```

File: src/operations/create_collection.ts [hunk 9]
```diff
@@ -74,25 +89,20 @@ export class CreateCollectionOperation extends CommandOperation<
...
     const cmd: Document = { create: name };
-    for (const n in options) {
-      if (
-        (options as any)[n] != null &&
-        typeof (options as any)[n] !== 'function' &&
-        !ILLEGAL_COMMAND_FIELDS.has(n)
-      ) {
-        cmd[n] = (options as any)[n];
+    for (const n in this) {
+      if (optionsAllowList.has(n) && this[n] != null && typeof this[n] !== 'function') {
```
Comment (emadum) [inline src/operations/create_collection.ts:104]:
```suggestion
      if (OPTIONS_ALLOW_LIST..has(n) && this[n] != null && typeof this[n] !== 'function') {
```

File: src/operations/indexes.ts [hunk 10]
```diff
@@ -189,18 +187,18 @@ export class CreateIndexesOperation extends CommandOperation<CreateIndexesOption
...
         );
         return;
       }
-      cmd.commitQuorum = options.commitQuorum;
+      cmd.commitQuorum = this.commitQuorum;
     }
 
     // collation is set on each index, it should not be defined at the root
-    this.options.collation = undefined;
+    // this.collation = undefined; FIXME!!!!!!
```
Comment (emadum) [inline src/operations/indexes.ts:201]: What's going on here?

File: src/operations/operation.ts [hunk 11]
```diff
@@ -64,20 +83,8 @@ export abstract class OperationBase<
...
-    // NOTE: Using the bang operator here because we know there is always a
-    //       session, explicit or implicit. We should disambiguate the session
-    //       from the options and set it as an explicit field
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    return this.options.session!;
-  }
-
   clearSession(): void {
-    delete this.options.session;
+    delete this.session;
```
Comment (emadum) [inline src/operations/operation.ts:87]: Echoing @reggi's comments, do want to use `delete` here or should we set it to `undefined`?

File: src/operations/operation.ts [hunk 12]
```diff
@@ -34,23 +34,42 @@ export interface OperationOptions extends BSONSerializeOptions {
...
+    }
+
+    return bsonOptions;
+  }
 
-  constructor(options: T = {} as T) {
-    this.options = Object.assign({}, options);
-    this.readPreference = ReadPreference.primary;
+  constructor(options = {}) {
+    // this.readPreference = ReadPreference.primary;
```
Comment (emadum) [inline src/operations/operation.ts:71]: This looks like it needs to be cleaned up, was this a temporary measure or do we want to delete this line?

File: src/operations/replace_one.ts (hunk 6 above)
Comment (emadum) [reply src/operations/replace_one.ts:22]: 🤔

File: test/functional/insert.test.js [hunk 13]
```diff
@@ -1891,9 +1889,14 @@ describe('Insert', function () {
...
-                var doc = docs.pop();
 
-                test.ok(doc.a._bsontype === 'Long');
+                docs.forEach((d, i) => {
+                  expect(d.a, `Failed on the document at index ${i}`).to.not.be.a('number');
+                  expect(d.a, `Failed on the document at index ${i}`).to.have.property('_bsontype');
+                  expect(d.a._bsontype, `Failed on the document at index ${i}`).to.be.equal('Long');
+                });
+
+                // expect(doc.a).to.have.property('_bsontype').to.be.equal('Long');
```
Comment (emadum) [inline test/functional/insert.test.js:1899]: Can you explain why these changes were necessary? Is this still testing the same thing?

**Review by emadum — CHANGES_REQUESTED** [review]
LGTM, this looks great. 🎉 Just a few nits/questions.

File: src/operations/indexes.ts (hunk 10 above)
Comment (emadum) [reply src/operations/indexes.ts:201]: The `Successfully create index with collation` test is failing in CI, probably related to this?

File: src/bulk/common.ts [hunk 14]
```diff
@@ -640,7 +640,7 @@ function executeCommands(
     err.ok = 0;
     // Merge top level error and return
     mergeBatchResults(batch, bulkOperation.s.bulkResult, err, undefined);
-    callback();
```
Comment (emadum) [inline src/bulk/common.ts:643]: This seems like it was probably an oversight, why would we want to swallow the error here?

File: test/functional/insert.test.js (hunk 13 above)
Comment (nbbeeken) [reply test/functional/insert.test.js:1899]: The bsonOptions weren't being inherited properly by the cursor, the first document would respect the options because it received them from the options parameter in the cursor's constructor. But subsequent getMores would get the bsonOptions off the operation 

https://github.com/mongodb/node-mongodb-native/blob/master/src/cursor/core_cursor.ts#L588

and this was just undefined and so we would fill in defaults that were different, so the following documents would have different deserialization. 

Anyway I changed the test to check that all of the documents `a` properties remain Long class objects rather than number primitives instead of just the first.

File: src/operations/operation.ts (hunk 12 above)
Comment (nbbeeken) [reply src/operations/operation.ts]: Cleaned up the readPreference was previously "overwritten" if the options object specified it, but now it will be a property directly on the class.

File: src/operations/indexes.ts (hunk 10 above)
Comment (nbbeeken) [reply src/operations/indexes.ts]: Yep fixed this, this was actually some cruft from TS conversion worth mentioning to spread the knowledge about it (@reggi), we used this syntax in few places:
```ts
class A {
  prop!: string;
}
```
because at the time the errors surrounding prop being optional were numerous and it felt right to assert the prop's presence, I just pushed changing props on `CommandOperation` that were asserted to be optional and there are no more type errors surrounding that. So we removed the dependence on the properties with further work that followed the big operations layer conversion.

File: src/operations/replace_one.ts (hunk 6 above)
Comment (nbbeeken) [reply src/operations/replace_one.ts]: Git blame says I left the FIXME there thats why I left 'neal why?' lol oops! looks like the option doesn't need to be there, it isn't a replaceOne option but it is an update option, maybe during typing this interface was being shared.

File: src/operations/map_reduce.ts (hunk 4 above)
Comment (nbbeeken) [reply src/operations/map_reduce.ts]: Is there more ops than mapReduce that do this? I think MapReduce does this since the operation takes in functions it makes sense to double check the options don't have any (for example, `validator` sounds like a function but is a document). I can get more context on this if there's more
